doc/docs/spec/typeclasses/index.md

```diff
@@ -0,0 +1,124 @@
+    function fndecl() -> int;               // 2
+
+    function fndef(arg : int) -> int {      // 3
+        return arg;
+    }
+};
+
+default instance tc(int) {                  // 4
+    function fndecl() -> int {              // 5
+        return 0;
+    }
+};
+
+let inst = instance tc(int) {               // 6
+    function fndecl() -> int {              // 7
+        return 1;
+    }
+
+    function fndef(inst_arg : int) -> int { // 8
+        return arg - 1;
+    }
+};
+```
+
+### Basics of typeclasses and instances
+
+ 1. `tc` is a name of a typeclass. The type of `tc` is `typeclass`, which is a builtin type. All typeclasses are of type
+    `typeclass`.
+ 2. `inst` is a name of a typeclass instance. The type of `inst` is `tc`.
+ 3. The type of the unnamed default instance is also `tc`.
+ 4. Values of type that is a typeclass can be used as a type. **Drafting note: this is somewhat unimportant for the time
+    being, at least until Vapor gets templates and/or dependent types, but shall be implemented from the start. In essence,
+    this allows to pass specific typeclass instances by-value by wrapping a value in them, but the exact specification of
+    that doesn't matter because it won't be useful until it can be inferred or passed in in any way.**
+ 5. `fndecl` declared on line (2) is a declaration of a typeclass function. This function _must_ be defined by an instance
+    for the instance to be valid.
+ 6. `fndecl`s defined on lines (5) and (7) are implementations of the function declared on line (2).
+ 7. `fndef` defined on line (3) is both a declaration of a typeclass function, and a default definition of such function.
+    If any instance of `tc` doesn't override this implementation, it is going to be inherited from the typeclass definition.
+    In other words, the default instance defined on line (4) uses this definition for its `fndef`.
+ 8. `fndef` defined on line (8) is a definition of the typeclass function declared on line (3) that overrides the default
+    implementation. The instance `inst` uses this definition, and not the default one, as the definition of the typeclass
+    function declared on line (3).
+
+### Access to typeclass and instance members
+
+A typeclass can be invoked with an argument to access the default instance for the given argument:
+
+```
+let default_instance = tc(int);
+```
+
+`default_instance` is of type `tc` and is equivalent to the default instance of `tc`.
+
+When a call is made through a typeclass instance, the definition of a function belonging to that instance is looked up.
+
+```
+tc(int).fndecl()                    // returns 0, because the definition (5) is used
+default_instance.fndecl()           // equivalent to the above
+inst.fndecl()                       // returns 1, uses definition (7)
+tc(int).fndef(123)                  // returns 123, per definition (3)
+inst.fndef(123)                     // returns 122, per definition (8)
+```
+
```

Comment:
I think it'd be worth mentioning what happens when you refer to an incomplete typeclass. That is, in the case of `tc`, it only has a default for int. So what if I refer to `tc(tc(int))` or `tc(float)` when there is no implementation for `.fndecl`?

Reply:
Compilation error when trying to say `tc(float)` because the instance doesn't exist. This also reminds me I haven't actually described the semantics of `implicit` typeclasses...

Reply:
Ah, and of course this is different with the default instance existing. Will write this down.